doc/internal/man3/OSSL_EVENT.pod

```diff
@@ -0,0 +1,217 @@
+ void ossl_event_free(OSSL_EVENT *event);
+
+ uint32_t ossl_event_get_type(const OSSL_EVENT *event);
+ uint32_t ossl_event_get_priority(const OSSL_EVENT *event);
+ OSSL_TIME ossl_event_get_when(const OSSL_EVENT *event);
+ void *ossl_event_get0_payload(const OSSL_EVENT *event, size_t *length);
+
+ int ossl_event_queue_schedule(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event);
+ int ossl_event_queue_delete(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event);
+
+ OSSL_TIME ossl_event_queue_time_until(OSSL_EVENT *event);
+ OSSL_TIME ossl_event_queue_time_until_next(const OSSL_EVENT_QUEUE *queue);
+
+ int ossl_event_queue_postpone(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event,
+                               OSSL_TIME how_long);
+ int ossl_event_queue_postpone_until(OSSL_EVENT_QUEUE *queue,
+                                     OSSL_EVENT *event,
+                                     OSSL_TIME when);
+
+ int ossl_event_queue_dispatch(OSSL_EVENT_QUEUE *queue);
+
+ OSSL_EVENT_SUBSCRIPTION *ossl_event_queue_subscribe(OSSL_EVENT_QUEUE *queue,
+                                                     uint32_t type,
+                                                     ossl_event_callback_fn *cb,
+                                                     void *cbarg);
+ void ossl_event_queue_unsubscribe(OSSL_EVENT_SUBSCRIPTION *sub);
+
+=head1 DESCRIPTION
+
+These functions implement an event queue.
+
+=head2 Event
+
+An event is a small structure, B<OSSL_EVENT>, carrying information:
+
+=over 4
+
+=item type
+
+A mandatory event type, which is a simple numeric identity, the
+meaning of which is not known by the event functionality itself.
+
+=item priority
+
+A manatory nonnegative integral quantity.  The lower the priority the earlier
+the event will be processes.
+
+=item when
+
+An optional time indicating when the event could be triggered.  Events are
+guaranteed to not trigger before their time.
+
+=item context
+
+A reference to user supplied contextual informaton.  The event queue passes
+this to callbacks and never dereferences the pointer.
+
+=item payload, payload_size
+
+A reference to some event specific data of a specified length.
+
+=back
+
+The event itself is designed for a single synchronous thread, i.e. cannot be
+shared by multiple threads.  The diverse objects it refers to may, however,
+be shared by multiple threads, at the discretion of the functions in the
+method structure.
+
+Once populated, the event type, the references to event context, and the
+reference to the destructor function are considered immutable, up until the
+event structure is destroyed.
+
+The reference to the auxilliary identifying material or to the payload,
+however, are considered mutable.  Any event handler may "steal" them and
+replace the reference to them in the event structure with NULL.  Stealing
+must be done with much care.
+
+Events may be embedded in another structure, as a static variable, or
+as an local variable in a function.  Events may also be dynamically allocated.
```

Comment:
Honestly I cannot imagine how static or local variable events would work. Perhaps immediate events could work for static ones but local variables that can go out of scope any time? Especially if scheduling and dispatch is separate.

Reply:
Also, if we want them embeddable, the structure must be public (internally only of course), do then the ossl_event_set and/or the getters make any sense?

Reply:
The big question is do we want them to be embeddable or not?
It's simplier to not but more efficient to allow it.

We're already got way too many mallocs happening :crying_cat_face: